components/brave_extension/extension/brave_extension/content_element_picker.ts

```diff
@@ -0,0 +1,424 @@
+}
+
+interface Rule {
+  type: Selector
+  value: any
+}
+
+class ElementSelectorBuilder {
+  public hasId: boolean
+  private elem: Element
+  private rules: Rule[]
+  private tag: string
+
+  constructor (elem: Element) {
+    this.rules = []
+    this.tag = ''
+    this.elem = elem
+    this.hasId = false
+  }
+
+  addRule (rule: Rule): void {
+    if (rule.type < Selector.Id || rule.type > Selector.NthOfType) {
+      console.log(`Unexpected selector: ${rule.type}`)
+      return
+    }
+    if (Array.isArray(rule.value) && rule.value.length === 0) {
+      return
+    }
+    if (rule.type === Selector.Id) { this.hasId = true }
+    this.rules.push(rule)
+  }
+
+  addTag (tag: string): void {
+    this.tag = tag
+  }
+
+  size (): number {
+    return this.rules.length
+  }
+
+  toString (mask: number = mostSpecificMask): string {
+    let selector = this.tag + ''
+    for (const rule of this.rules) {
+
+      if (!(mask & SpecificityFlags.Id) && rule.type === Selector.Id) {
+        continue
+      }
+      if (!(mask & SpecificityFlags.Class) && rule.type === Selector.Class) {
+        continue
+      }
+      if (!(mask & SpecificityFlags.Attributes) && rule.type === Selector.Attributes) {
+        continue
+      }
+      if (!(mask & SpecificityFlags.NthOfType) && rule.type === Selector.NthOfType) {
+        continue
+      }
+      if (this.hasId && (mask & SpecificityFlags.Id) && rule.type === Selector.Class) {
+        continue
+      }
+
+      switch (rule.type) {
+        case Selector.Id: {
+          selector += '#' + rule.value
+          break
+        }
+        case Selector.Class: {
+          selector += '.' + rule.value.join('.')
+          break
+        }
+        case Selector.Attributes: {
+          for (const attribute of rule.value) {
+            const sourceAttr = this.elem.getAttribute(attribute.attr)
+            let op = '*='
+            if (attribute.attr === sourceAttr) {
+              op = '='
+            } else if (attribute.attr.startsWith(sourceAttr)) {
+              op = '^='
+            }
+            selector += `[${attribute.attr}${op}"${attribute.value}"]`
+          }
+          break
+        }
+        case Selector.NthOfType: {
+          // NOTE: this selector is not valid without calling addTag(..)
+          selector += `:nth-of-type(${rule.value})`
+          break
+        }
+        default: { /* Unreachable */ }
+      }
+    }
+    return selector
+  }
+}
+
+// We search for a CSS selector for the target element. We want the most specific identifiers.
+const cssSelectorFromElement = (elem: Element): ElementSelectorBuilder => {
+  const builder = new ElementSelectorBuilder(elem)
+
+  // ID
+  if (elem.id.length > 0) {
+    builder.addRule({
+      type: Selector.Id,
+      value: elem.id
+    })
+  }
+
+  // Class names
+  if (elem.classList.length > 0) {
+    builder.addRule({
+      type: Selector.Class,
+      value: [...elem.classList].map((c: string) => CSS.escape(c))
```

Comment:
Why escape the class names here?

Reply:
Also, will this generate false positives on the page, if there are, say 10 list elements w/ the same "list-item" class, and the user only wants to block one fo them?  (is this dealt with elsewhere?(

Also, if there is an id case here, why also have the class name case?  Should each of these if cases have an early return?

Reply:
we use the class names to build a query selector rule, so some added safety safe

Reply:
as for your second comment, we exclude the class if we match the id here. we could also move that logic to the selector building phase too, if you think that's better

https://github.com/brave/brave-core/blob/5b9daec5474a6af71633b811c16256a8ad66154e/components/brave_extension/extension/brave_extension/content_element_picker.ts#L98